### ISSUE TITLE
searcher() does not work properly when handling multiple requests to obtain documents

### ISSUE THREAD

**Issue by ThomasWaldmann** [moinwiki/moin#480]
**[Original report](https://bitbucket.org/thomaswaldmann/moin-2.0/issue/480) by johnyoon (Bitbucket: [johnyoon](https://bitbucket.org/johnyoon), GitHub: [johnyoon](https://github.com/johnyoon)).**

----------------------------------------

The prerequisite to this issue is #479 where you first have to make the `_modify_index` method to have BufferedWriter.  

./MoinMoin/storage/middleware/indexing.py

```
Traceback (most recent call last):                                                                                           
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app                    
    response = self.full_dispatch_request()                                                                                  
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request       
    rv = self.handle_user_exception(e)                                                                                       
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception       
    reraise(exc_type, exc_value, tb)                                                                                         
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request       
    rv = self.dispatch_request()                                                                                             
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request            
    return self.view_functions[rule.endpoint](**req.view_args)                                                               
  File "/srv/www/moin/MoinMoin/apps/frontend/views.py", line 636, in modify_item                                             
    return item.do_modify()                                                                                                  
  File "/srv/www/moin/MoinMoin/items/__init__.py", line 876, in do_modify                                                    
    self.modify(meta, data, comment, contenttype_guessed, **{CONTENTTYPE: contenttype_qs})                                   
  File "/srv/www/moin/MoinMoin/items/__init__.py", line 493, in modify                                                       
    return self._save(meta, data, contenttype_guessed=contenttype_guessed, comment=comment)                                  
  File "/srv/www/moin/MoinMoin/items/__init__.py", line 635, in _save                                                        
    return_rev=True,                                                                                                         
  File "/srv/www/moin/MoinMoin/storage/middleware/protecting.py", line 314, in store_revision                                
    rev = self.item.store_revision(meta, data, overwrite=overwrite, return_rev=return_rev, fqname=fqname, **kw)              
  File "/srv/www/moin/MoinMoin/storage/middleware/indexing.py", line 1137, in store_revision                                 
    return Revision(self, revid)                                                                                             
  File "/srv/www/moin/MoinMoin/storage/middleware/indexing.py", line 1178, in __init__                                       
    raise KeyError                                                                                                           
KeyError                                                                                                                     
```

From the documentation, there is a way to get the latest searcher by using `refresh()`
https://pythonhosted.org/Whoosh/threads.html

```
#!/python
 def _document(self, idx_name=LATEST_REVS, **kw):
     """
     Return a document matching the kw args (internal use only).
     """
     logging.info(idx_name)
     logging.info(self.ix[idx_name])
     with self.ix[idx_name].searcher().refresh() as searcher:
         return searcher.document(**kw)
```

but I do not think it is still fast enough, because it still fails to obtain the `doc` with the same error or the following:

```
#!/bash

Traceback (most recent call last):
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/srv/www/moin-venv-python/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/srv/www/moin/MoinMoin/apps/frontend/views.py", line 636, in modify_item
    return item.do_modify()
  File "/srv/www/moin/MoinMoin/items/__init__.py", line 876, in do_modify
    self.modify(meta, data, comment, contenttype_guessed, **{CONTENTTYPE: contenttype_qs})
  File "/srv/www/moin/MoinMoin/items/__init__.py", line 493, in modify
    return self._save(meta, data, contenttype_guessed=contenttype_guessed, comment=comment)
  File "/srv/www/moin/MoinMoin/items/__init__.py", line 635, in _save
    return_rev=True,
  File "/srv/www/moin/MoinMoin/storage/middleware/protecting.py", line 317, in store_revision
    return ProtectedRevision(self.protector, rev, p_item=self)
  File "/srv/www/moin/MoinMoin/storage/middleware/protecting.py", line 343, in __init__
    self.item = p_item or ProtectedItem(protector, rev.item)
  File "/srv/www/moin/MoinMoin/storage/middleware/protecting.py", line 246, in __nonzero__
    return bool(self.item)
  File "/srv/www/moin/MoinMoin/storage/middleware/indexing.py", line 1011, in __nonzero__
    return self.itemid is not None
  File "/srv/www/moin/MoinMoin/storage/middleware/indexing.py", line 963, in _get_itemid
    return self._current.get(ITEMID)

```

**Comment by RogerHaase** [comment]
Please add instructions to reproduce this error. What did you use to create the stress test?

**Comment by RogerHaase** [comment]
fixed by cc576d6
